File: prep/implementation/main_iterations/iteration_cpp/kruskal.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class DSUF {
private:
  vector<int> rank;
  vector<int> parent;

  void make_set(int p) {
    this->rank.push_back(0);
    this->parent.push_back(p);
  }

public:
  DSUF(int sz) {
    for (int i = 0; i < sz; i++) {
      make_set(i);
    }
  }

  int dsuf_find(int p) {
    if (p != this->parent[p]) {
      p = dsuf_find(this->parent[p]);
    }

    return this->parent[p];
  }

  void dsuf_union(int p, int q) {
    int pid = dsuf_find(p);
    int qid = dsuf_find(q);

    if (pid == qid) {
      return;
    }

    if (this->rank[pid] > this->rank[qid]) {
      this->parent[qid] = pid;
    } else {
      this->parent[pid] = qid;
      if (this->rank[pid] == this->rank[qid]) {
        this->rank[qid]++;
      }
    }
  }
};

struct Edge {
  int from;
  int to;
  int weight;

  Edge(int from, int to, int weight) {
    this->from = from;
    this->to = to;
    this->weight = weight;
  }
};

void kruskal(DSUF &dsuf, vector<Edge> &edges) {
  sort(edges.begin(), edges.end(),
       [](const auto &e1, const auto &e2) { return e1.weight < e2.weight; });

  vector<Edge> mst_edges;
  int mst_cost = 0;

  for (Edge edge : edges) {
    int from = edge.from;
    int to = edge.to;
    int weight = edge.weight;

    if (dsuf.dsuf_find(from) != dsuf.dsuf_find(to)) {
      dsuf.dsuf_union(from, to);
      mst_cost += weight;
      mst_edges.push_back(edge);
    }
  }

  cout << mst_cost << "\n";
  for (Edge edge : mst_edges) {
    cout << edge.from << " " << edge.to << "\n";
  }
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, m, from, to, weight;
  cin >> n >> m;

  DSUF dsuf(n);
  vector<Edge> edges;
  for (int i = 0; i < m; i++) {
    cin >> from >> to >> weight;
    edges.push_back(Edge(from, to, weight));
  }

  kruskal(dsuf, edges);

  return 0;
}

File: lc/coreprep/implementation/main_reference_implementation/Kruskal.java
import java.util.*;

public class Kruskal {
  static class DSUF {
    private int[] rank;
    private int[] parent;

    DSUF(int size) {
      this.rank = new int[size];
      this.parent = new int[size];

      for (int i = 0; i < size; i++) {
        makeSet(i);
      }
    }

    private void makeSet(int p) {
      this.rank[p] = 0;
      this.parent[p] = p;
    }

    public int find(int p) {
      if (p != parent[p]) {
        parent[p] = find(parent[p]);
      }

      return parent[p];
    }

    public void union(int p, int q) {
      int pid = find(p);
      int qid = find(q);

      if (pid == qid) {
        return;
      }

      if (rank[pid] > rank[qid]) {
        parent[qid] = pid;
      } else {
        parent[pid] = qid;
        if (rank[pid] == rank[qid]) {
          rank[qid]++;
        }
      }
    }
  }

  static class Edge {
    int from;
    int to;
    int weight;

    Edge(int from, int to, int weight) {
      this.from = from;
      this.to = to;
      this.weight = weight;
    }

    @Override
    public boolean equals(Object o) {
      if (o == null || !(o instanceof Edge)) {
        return false;
      }

      Edge other = (Edge)o;
      return this.from == other.from && this.to == other.to ||
          this.from == other.to && this.to == other.from;
    }

    @Override
    public int hashCode() {
      return Objects.hash(this.from, this.to);
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      DSUF dsuf = new DSUF(n);

      int m = in.nextInt();
      List<Edge> edges = new ArrayList<>();
      for (int i = 0; i < m; i++) {
        int from = in.nextInt();
        int to = in.nextInt();
        int weight = in.nextInt();

        edges.add(new Edge(from, to, weight));
      }

      kruskal(dsuf, edges);
    }
  }

  // O(|E| log|V|)
  private static void kruskal(DSUF dsuf, List<Edge> edges) {
    edges.sort((e1, e2) -> Integer.compare(e1.weight, e2.weight));

    List<Edge> mstEdges = new ArrayList<>();
    int mstCost = 0;

    for (Edge e : edges) {
      if (dsuf.find(e.from) != dsuf.find(e.to)) {
        mstEdges.add(e);
        dsuf.union(e.from, e.to);
        mstCost += e.weight;
      }
    }

    System.out.println(mstCost);
    for (Edge e : mstEdges) {
      System.out.printf("%d %d\n", e.from, e.to);
    }
  }
}